File: src/app_store_connect/api/apps.py
"""
Apps API module for App Store Connect
"""

from typing import Dict, Any, Optional, List
from ..base import BaseAPI


class AppsAPI(BaseAPI):
    """
    Manage apps in App Store Connect
    """
    
    def get_all(self, limit: int = 200) -> List[Dict[str, Any]]:
        """
        Get all apps
        
        Args:
            limit: Number of results per page
            
        Returns:
            List of app data
        """
        return self.get_all_pages('apps', limit=limit)
    
    def get(self, app_id: str) -> Dict[str, Any]:
        """
        Get a specific app
        
        Args:
            app_id: The app ID
            
        Returns:
            App data
        """
        response = super().get(f'apps/{app_id}')
        return response['data']
    
    def get_by_bundle_id(self, bundle_id: str) -> Optional[Dict[str, Any]]:
        """
        Get app by bundle ID
        
        Args:
            bundle_id: The bundle identifier
            
        Returns:
            App data or None if not found
        """
        response = super().get('apps', params={'filter[bundleId]': bundle_id})
        data = response.get('data', [])
        return data[0] if data else None
    
    def update(self, app_id: str, **attributes) -> Dict[str, Any]:
        """
        Update app attributes
        
        Args:
            app_id: The app ID
            **attributes: Attributes to update
            
        Returns:
            Updated app data
        """
        data = {
            'data': {
                'type': 'apps',
                'id': app_id,
                'attributes': attributes
            }
        }
        response = self.patch(f'apps/{app_id}', data=data)
        return response['data']
    
    def get_app_infos(self, app_id: str) -> List[Dict[str, Any]]:
        """
        Get app info records for an app
        
        Args:
            app_id: The app ID
            
        Returns:
            List of app info data
        """
        response = super().get(f'apps/{app_id}/appInfos')
        return response.get('data', [])
    
    def get_app_store_versions(self, app_id: str) -> List[Dict[str, Any]]:
        """
        Get app store versions for an app
        
        Args:
            app_id: The app ID
            
        Returns:
            List of app store version data
        """
        response = super().get(f'apps/{app_id}/appStoreVersions')
        return response.get('data', [])
    
    def get_builds(self, app_id: str) -> List[Dict[str, Any]]:
        """
        Get builds for an app
        
        Args:
            app_id: The app ID
            
        Returns:
            List of build data
        """
        response = super().get(f'apps/{app_id}/builds')
        return response.get('data', [])

File: rust/src/api/apps.rs
use crate::base::BaseAPI;
use crate::error::AppStoreConnectError;
use serde_json::{json, Value};
use std::collections::HashMap;

#[derive(Clone)]
pub struct AppsAPI {
    base: BaseAPI,
}

impl AppsAPI {
    pub fn new(base: BaseAPI) -> Self {
        Self { base }
    }

    pub async fn get_all(&self, limit: Option<u32>) -> Result<Vec<Value>, AppStoreConnectError> {
        self.base.get_all_pages("apps", None, limit).await
    }

    pub async fn get_app(&self, app_id: &str) -> Result<Value, AppStoreConnectError> {
        let response = self.base.get(&format!("apps/{}", app_id), None).await?;
        response
            .get("data")
            .cloned()
            .ok_or_else(|| AppStoreConnectError::Api {
                message: "Invalid response format".to_string(),
            })
    }

    pub async fn get_by_bundle_id(
        &self,
        bundle_id: &str,
    ) -> Result<Option<Value>, AppStoreConnectError> {
        let mut params = HashMap::new();
        params.insert("filter[bundleId]".to_string(), bundle_id.to_string());

        let response = self.base.get("apps", Some(params)).await?;

        if let Some(data) = response.get("data").and_then(|d| d.as_array()) {
            Ok(data.first().cloned())
        } else {
            Ok(None)
        }
    }

    pub async fn update(
        &self,
        app_id: &str,
        attributes: Value,
    ) -> Result<Value, AppStoreConnectError> {
        let data = json!({
            "data": {
                "type": "apps",
                "id": app_id,
                "attributes": attributes
            }
        });

        let response = self.base.patch(&format!("apps/{}", app_id), data).await?;
        response
            .get("data")
            .cloned()
            .ok_or_else(|| AppStoreConnectError::Api {
                message: "Invalid response format".to_string(),
            })
    }

    pub async fn get_app_infos(&self, app_id: &str) -> Result<Vec<Value>, AppStoreConnectError> {
        let response = self
            .base
            .get(&format!("apps/{}/appInfos", app_id), None)
            .await?;

        if let Some(data) = response.get("data").and_then(|d| d.as_array()) {
            Ok(data.clone())
        } else {
            Ok(Vec::new())
        }
    }

    pub async fn get_app_store_versions(
        &self,
        app_id: &str,
    ) -> Result<Vec<Value>, AppStoreConnectError> {
        let response = self
            .base
            .get(&format!("apps/{}/appStoreVersions", app_id), None)
            .await?;

        if let Some(data) = response.get("data").and_then(|d| d.as_array()) {
            Ok(data.clone())
        } else {
            Ok(Vec::new())
        }
    }

    pub async fn get_builds(&self, app_id: &str) -> Result<Vec<Value>, AppStoreConnectError> {
        let response = self
            .base
            .get(&format!("apps/{}/builds", app_id), None)
            .await?;

        if let Some(data) = response.get("data").and_then(|d| d.as_array()) {
            Ok(data.clone())
        } else {
            Ok(Vec::new())
        }
    }
}
